### ISSUE TITLE
The main thread is blocked when `DescriptorRunnerWebGL` is running

### ISSUE THREAD

**Issue by zhangjk95** [mil-tokyo/webdnn#530]
I noticed that `DescriptorRunnerWebGL.run` blocks the main thread although it is tagged `async`. The web app does not respond to any events while `DescriptorRunnerWebGL` is running. Is there any solution to solve this problem? (For example, use Web Workers?)

**Comment by Kiikurage** [comment]
When downloading results from GPU to CPU, [WebGLRenderingContext.readPixels()
](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/readPixels) blocks main thread until all dispatched GPU programs finished. 

As you said, using worker is one solution, but currently almost of all browsers doesn't support workers with WebGL rendering context ([OffscreenCanvas](https://developer.mozilla.org/ja/docs/Web/API/OffscreenCanvas)).

Another solution is separate execution method into dispatching and downloading. With using this approach, while calculation, main thread is not blocked.

```
Before)    ----(time)--->

        GPU                     |======Kernel 1======|======Kernel 2======|
                                :                                         :
     Runner |=dispatch Kernel 1=|=dispatch Kernel 2=|.. wait until done ..| download all results |
            :                                                                                    :
Main Thread |===============================BLOCKED==============================================|
```

```
After) 

        GPU                     |======Kernel 1======|======Kernel 2======|
                                :
     Runner |=dispatch Kernel 1=|=dispatch Kernel 2=|                   | download all results |
            :                                       :                   :                      :
Main Thread |==============BLOCKED==================|                   |========BLOCKED=======|
```

However, there is no method to know when the computation is done asynchronously. Therefore this approach cannot be used right now.

If you have any idea, please share with us. 

Thank you.

**Comment by zhangjk95** [comment]
Ok. I got your point.
After some investigation, I found a function in WebGL2 called `clientWaitSync` that can judge whether the calculation is completed or not. So it is possible to wait for the computation be to done by calling this function periodically.

e.g:
```TypeScript
async waitForComplete() {
    let gl = this.handler.gl as any;
    let sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
    let status = gl.clientWaitSync(sync, 0, 0);

    while (status !== gl.CONDITION_SATISFIED && status !== gl.ALREADY_SIGNALED) {
        await this.promiseTimeout(1);
        status = gl.clientWaitSync(sync, 0, 0);
    }

    gl.deleteSync(sync);
}

promiseTimeout(timeout) {
    return new Promise((resolve, reject) => {
        setTimeout(resolve, timeout);
    });
}
```

The function `waitForComplete` above is somehow equivalent to `gl.finish()`, but it can run asynchronously .

However, using this function is a BAD idea, because most of the kernels can finish calculation in less than 1 ms, but the callback of `setTimeout`/`setInterval` can only run once every 4 ms.
 ([https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Timeouts_throttled_to_>4ms](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Timeouts_throttled_to_>4ms)) Using this function every time after calling `gl.drawArrays` will cause the the `DescriptorRunnerWebGL` to run much slower than before. (I tested in Chrome with MGM model, using this function makes it 2-3 times slower)

Therefore, I don't have any good idea yet.
